### PR TITLE
ORTTrainer / ORTSeq2SeqTrainer refactoring - Inference with ORTModel

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

This PR replaces the bare inference sessions in `ORTTrainer` with subclasses of `ORTModel`, and enables the inference with ONNX Runtime in `ORTSeq2SeqTrainer`.

__[Refactoring `ORTTrainer`]__
* Replace inference session in `evaluation_loop_ort` and `prediction_loop_ort` with subclasses of `ORTModel`.
* Clean up unused in `evaluation_loop_ort` and `prediction_loop_ort`

__[Refactoring `ORTSeq2SeqTrainer`]__

* Adapt `OnnxConfigWithPastAndLoss` on the top of `DecoderOnnxConfig`.
* Generate dummy labels and compute of loss in `_DecoderWithLMhead`.
* Add labels and the loss in sequence-to-sequence ort modeling.
* Override `evaluation_loop_ort` and `prediction_loop_ort` to leverage `ORTModelForSeq2SeqLM`.

### CODE REVIEW

**Review by regisss — APPROVED** [review]
It looks great Jingya!

File: optimum/onnxruntime/trainer.py [hunk 1]
```diff
@@ -764,7 +762,7 @@ def evaluation_loop_ort(
         Works both with or without labels.
         """
         logger.info("[INFO] ONNX Runtime inference starts...")
-        self.infer_sess = None
+        self.ort_model = None
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:765]: Is it necessary for `ort_model` to be an attribute?

**Review by echarlaix — APPROVED** [review]
LGTM 🔥

File: optimum/onnxruntime/trainer.py (hunk 1 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:765]: Nope, it would be enough to take it as a function variable. Good point, will correct this.

File: optimum/onnxruntime/trainer.py [hunk 2]
```diff
@@ -788,45 +814,28 @@ def evaluation_loop_ort(
...
-
-        # if full fp16 or bf16 eval is wanted and this ``evaluation`` or ``predict`` isn't called
-        # while ``train`` is running, cast it to the right dtype first and then put on device
-        if not self.is_in_train:
-            if args.fp16_full_eval:
-                model = model.to(dtype=torch.float16, device=args.device)
-            elif args.bf16_full_eval:
-                model = model.to(dtype=torch.bfloat16, device=args.device)
-
-        batch_size = self.args.eval_batch_size
```
Comment (regisss) [inline optimum/onnxruntime/trainer.py:829]: This code snippet is not needed anymore?

File: optimum/onnxruntime/trainer.py [hunk 3]
```diff
@@ -835,8 +844,6 @@ def evaluation_loop_ort(
             logger.info("  Num examples: Unknown")
         logger.info(f"  Batch size = {batch_size}")
 
-        model.eval()
```
Comment (regisss) [inline optimum/onnxruntime/trainer.py:838]: Is this done elsewhere?

**Review by regisss — COMMENTED** [review]
Great work Jingya :fire: 
I just have two naive questions

File: optimum/onnxruntime/trainer.py (hunk 2 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:829]: Hi @regisss,
In the code beyond, for the deepspeed part, as ONNX Runtime doesn't support DeepSpeed stage 3 / infinity yet, it is not necessary to initialize the deepspeed engine during inference. And as for the model weights dtype conversion, as for the inference the model is an ORTModel with an ORT inference session behind, it won't work this way. 
Thanks for pointing that out, indeed currently there are some issues on dtype conversion especially for models trained with mixed-precision. I will add that in the backlog.

Comment (regisss) [reply optimum/onnxruntime/trainer.py:829]: Thanks for the explanation Jingya :slightly_smiling_face:

File: optimum/onnxruntime/trainer.py (hunk 3 above)
Comment (regisss) [reply optimum/onnxruntime/trainer.py:838]: Ah okay I see, thanks!

**Review by regisss — APPROVED** [review]

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 4]
```diff
@@ -558,7 +563,11 @@ def forward(
         past_key_values = tuple(past_key_values[i : i + num_pkv] for i in range(0, len(past_key_values), num_pkv))
         logits = torch.from_numpy(outputs[self.output_names["logits"]]).to(self._device)
 
-        return Seq2SeqLMOutput(logits=logits, past_key_values=past_key_values)
+        if labels is None:
+            return Seq2SeqLMOutput(logits=logits, past_key_values=past_key_values)
+        else:
+            loss = torch.from_numpy(outputs[self.output_names["loss"]]).to(self._device)
+            return Seq2SeqLMOutput(loss=loss, logits=logits, past_key_values=past_key_values)
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_seq2seq.py:570]:
```suggestion
        loss = None
        if "loss" in self.output_names:
            loss = torch.from_numpy(outputs[self.output_names["loss"]]).to(self._device)
        return Seq2SeqLMOutput(loss=loss, logits=logits, past_key_values=past_key_values)
```

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 5]
```diff
@@ -596,22 +606,28 @@ def forward(
...
+        decoder_outputs = self.decoder(**decoder_inputs)
+
+        if labels is None:
+            return Seq2SeqLMOutput(logits=decoder_outputs.logits, past_key_values=decoder_outputs.past_key_values)
         else:
-            decoder_outputs = self.decoder_with_past(
-                input_ids=decoder_input_ids[:, -1:],  # Cut decoder_input_ids if past is used
-                past_key_values=past_key_values,
-                encoder_hidden_states=encoder_outputs.last_hidden_state,
-                encoder_attention_mask=attention_mask,
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_seq2seq.py:610]: `self.decoder_with_past` should not be removed and should be used when `use_cache` is True and when `past_key_values` have been computed

**Review by echarlaix — CHANGES_REQUESTED** [review]

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 5 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_seq2seq.py:610]: Yes, you ar right. Putting it back!

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 6]
```diff
@@ -596,21 +626,29 @@ def forward(
             encoder_outputs = self.encoder(input_ids=input_ids, attention_mask=attention_mask)
 
         # Decode
+        decoder_inputs = {
+            "encoder_hidden_states": encoder_outputs.last_hidden_state,
+            "encoder_attention_mask": attention_mask,
+        }
```
Comment (echarlaix) [inline optimum/onnxruntime/modeling_seq2seq.py:632]: I would be in favor of removing the `decoder_inputs` dictionnary and just add the labels arguments to both the `decoder` / `decoder_with_past`

Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_seq2seq.py]: I get what you mean, will put the original inference back.

**Review by echarlaix — APPROVED** [review]
LGTM 🔥
